Add scikit-learn training and serving pipeline

luotigerlsx opened the pull request:
This is a example for building a scikit-learn-based machine learning pipeline trainer that can be run on AI Platform, which is built on top of the [scikit-learn template](https://github.com/GoogleCloudPlatform/cloudml-samples/tree/master/sklearn/sklearn-template/template). The pipeline can be trained locally or remotely on AI platform. The trained model can be further deployed on AI platform to serve online traffic. The entire pipeline includes three major components:

1. Transformer: generate new features from raw_data with user defined logic (function).
2. Pre-processer: handle typical standard pre-processing e.g. scaling, imputation, one-hot-encoding and etc.
3. Estimator: exact machine learning model e.g. RandomForestClassifier.

Compared with the [scikit-learn template](https://github.com/GoogleCloudPlatform/cloudml-samples/tree/master/sklearn/sklearn-template/template), this example has the following additional feature:

1. Support both Classification and Regression, which can be specified in the configuration
2. Support serving for both JSON and List of Value formats
3. Support additional custome transformation logics besides typical pre-processing provided by scikit-learn

luotigerlsx commented:
This PR has also been reviewed by @cfezequiel during development.

janpeuker commented on examples/cloudml-sklearn-pipeline/.pylintrc:
```diff
@@ -0,0 +1,334 @@
+[MASTER]
```
I am not sure we should submit dot files, in particular pylint - the contributor guidelines already refer to the Python Style Guide

luotigerlsx replied:
Since this will be eventually used as standalone template or tool, i think it is still make sense to have it.

janpeuker commented on examples/cloudml-sklearn-pipeline/README.md:
```diff
@@ -0,0 +1,430 @@
+### Step 1. Tailor the scikit-learn trainer to your data
+`metadata.py` is where the dataset's metadata is defined. 
+By default, the file is configured to train on the Census dataset, which can be found at 
+`bigquery-public-data.ml_datasets.census_adult_income`.
```
Link to BigQuery Public Datasets website?

luotigerlsx replied:
Added as suggested.

janpeuker commented on examples/cloudml-sklearn-pipeline/scripts/deploy.sh:
```diff
@@ -0,0 +1,73 @@
+REGION=us-central1
+FRAMEWORK=SCIKIT_LEARN
+RUN_TIME=1.13
+PYTHON_VERSION=3.5 # only support python 2.7 and 3.5
```
A few times throughout the doc Python 3.5 is mentioned, I would make clear throughout which version is used instead of mentioning options

luotigerlsx replied:
Both version are supported actually. I have added explanation in the readme to make it clearer as "Note that both Python 2.7 and Python 3.5 are supported, but Python 3.5 is the recommended one since 2.7 is [deprecated](https://pythonclock.org/) soon"

janpeuker replied:
Can you check this is the same across all files, e.g. in the config Yaml I still see a different convention. What I meant was across all files (including Readme) make clear what version you run.

luotigerlsx replied:
Thanks for the comment. I have checked all, and ensure the message is consistent.

smhosein commented on examples/cloudml-sklearn-pipeline/README.md:
```diff
@@ -0,0 +1,430 @@
+## Pipeline overview
+The overall flow of the pipeline can be summarized as follows and illustrated in the flowchart:
+
+**Raw Data -> Transformer -> Pre-processor -> Estomator -> Trained Pipeline**
```
Spelling: Estimator

luotigerlsx replied:
Good catch :). Modified.

smhosein commented on examples/cloudml-sklearn-pipeline/config/config.yaml:
```diff
@@ -0,0 +1,4 @@
+trainingInput:
+  scaleTier: STANDARD_1
+  runtimeVersion: "1.13"
+  pythonVersion: "3.5"  # Note: Python 3 is also supported
```
Since you are using python 3, I do not think you need this note in the files

luotigerlsx replied:
I have modified as "# Note: Both Python 2.7 and Python 3 is also supported"

smhosein commented on examples/cloudml-sklearn-pipeline/setup.py:
```diff
@@ -0,0 +1,39 @@
+
+setuptools.setup(
+    name='custom_scikit_learn',
+    author='luoshixin@google.com',
```
Maybe not put your email?

luotigerlsx replied:
Sure, I have removed the email, and replaced it with simply the name.

luotigerlsx commented:
@smhosein thanks a lot for the reviewing. I have incorporated your comments.
@janpeuker I have also addressed you comments, would you mind take a more look. Thanks.

janpeuker commented on examples/cloudml-sklearn-pipeline/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+matplotlib>=2.2.3
+seaborn>=0.9.0
+scikit-learn>=0.20.2
```
I suppose you've checked all of those licenses and attributed them correctly?

luotigerlsx replied:
Sure 👍

luotigerlsx commented:
Hey @jaketf, this PR has been reviewed. Please kindly help to merge it :)
